File: chaotic-queue/kotlin/chaoticQueue.kt

import kotlin.collections.*

fun swap(arr: Array<Int>, x: Int, y: Int) {
    val pair = Pair(arr[x], arr[y])
    arr[x] = pair.second
    arr[y] = pair.first
}

fun minimumBribes(q: Array<Int>): Unit {
    var count = 0
    var invalid = false

    val arr = q.copyOf()

    for (i in arr.size downTo 1 step 1) {
        println("i: $i")
        val temp = arr[i - 1]
        println("arr[i - 1]: $temp")
        println("arr: ${arr.joinToString()}")
        if (i == arr[i - 1]) {
            println("no bribe occurred")
            // no bribe occurred, the right marker is
            // in the right place
        } else if (i > 1 && i == arr[i - 2]) {
            println("one bribe occurred")
            // one bribe occurred
            // undo it
            swap(arr, i - 1, i - 2)
            count++
        } else if (i > 2 && i == arr[i - 3]) {
            println("two bribes occurred")
            // two bribes occurred
            // undo them
            swap(arr, i - 2, i - 3)
            swap(arr, i - 1, i - 2)
            count += 2
        } else {
            // person originally in position i
            // committed more than 2 bribes
            invalid = true
        }
        
        println("\n")
    }

    if (invalid) println("Too chaotic")
    else println(count)
}

fun main() {
    minimumBribes(arrayOf(1, 2, 5, 3, 7, 8, 6, 4))
}

File: chaotic-queue/javascript/chaoticQueue.js
function swap(arr, x, y) {
    const array = Array(...arr)
    const temp = array[x]
    array[x] = array[y]
    array[y] = temp

    return array
}

// Complete the minimumBribes function below.
function minimumBribes(q) {
    let count = 0
    let invalid = false

    let arr = Array(...q)

    for (let i = arr.length; i > 0; i--) {
        if (i == arr[i - 1]) {
            // no bribe occurred, the right marker is
            // in the right place
        } else if (i == arr[i - 2]) {
            // one bribe occurred
            // undo it
            arr = swap(arr, i - 1, i - 2)
            count++
        } else if (i == arr[i - 3]) {
            // two bribes occurred
            // undo them
            arr = swap(arr, i - 2, i - 3)
            arr = swap(arr, i - 1, i - 2)
            count += 2
        } else {
            // person originally in position i
            // committed more than 2 bribes
            invalid = true
        }
    }

    if (invalid) console.log("Too chaotic")
    else console.log(count)
}

minimumBribes([2, 1, 5, 3, 4])
